### PR TITLE
Microservices Example

### PR DESCRIPTION
**Pull request by benstopford**
A small ecosystem of services that provides Orders Functionality over a CQRS-styled interface.  

**OrdersService**
- Accepts orders via REST
- Allows orders to be Queried, blocking if they are not present. (Uses non blocking IO and Queryable state).

**InventoryService / FraudService / OrderDetailsService** 
- Provide different validations of the order, run in parallel. 
- First two demo the KStreams API, on the transactional producer and consumer. 

**Validations Aggregator Service** 
- Aggregates validations together and updates the corresponding order

**TODO** 
- Demo embedded KSQL around the input of Inventory (which can be done without Avro support)
- Add multi-node Kafka support to the test framework. 
- Add tests that use KStreamTestDriver to demonstrate how that works.

### CODE REVIEW

File: src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            System.out.println("Failed to stop " + getClass().getSimpleName() + " in 1000ms");
+        }
+        System.out.println(getClass().getSimpleName() + " was stopped");
+    }
+
+    private boolean isValid(Order order) {
+        if (order.getCustomerId() == null) return false;
+        if (order.getQuantity() < 0) return false;
+        if (order.getPrice() < 0) return false;
+        if (order.getProduct() == null) return false;
```
Comment (dguy) [inline src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java]: `return order.getProduct() != null` ?

**Comment by benstopford** [comment]
Thanks @dguy. On the orderid being versioned, it could be either, but the advantage of having the version on the OrderId is there is no way for the processing of one version of an order to affect another. So say I get two order updates in quick succession

File: src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ * Validates the details of each order.
+ * - Is the quantity positive?
+ * - Is there a customerId
+ * - etc...
+ * <p>
+ * This service could be built with Kafka Streams but we've used a Producer/Consumer pair
+ * including the integration with Kafka's Exactly Once feature (Transactions) to demonstrate
+ * this other style of building event driven services.
+ */
+public class OrderDetailsService implements Service {
```
[the remark this answers is not in the document]
Comment (benstopford) [reply src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java]: just to show that you can really, but maybe that's sending the wrong message.

**Comment by benstopford** [comment]
@dguy I've addressed all your comments. Thank you for them. I'm going to leave the other TODO items in the description above for a follow up PR as I'd like to get this one merged. You ok to lgtm?

**Pull request merged**